leshan-server-core/src/main/java/org/eclipse/leshan/server/impl/LwM2mRequestSenderImpl.java

```diff
@@ -0,0 +1,170 @@
+import org.eclipse.leshan.server.client.ClientRegistry;
+import org.eclipse.leshan.server.queue.impl.QueuedRequestSender;
+import org.eclipse.leshan.server.request.LwM2mRequestSender;
+import org.eclipse.leshan.server.response.ResponseListener;
+import org.eclipse.leshan.server.response.ResponseProcessingTask;
+import org.eclipse.leshan.util.NamedThreadFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Client binding mode aware request sender. i.e: provides a wrapper delegating the actual invocation to either
+ * {@link QueuedRequestSender} or a default {@link LwM2mRequestSender} depending upon whether the client has connected
+ * in Queue mode or not.
+ */
+public class LwM2mRequestSenderImpl implements LwM2mRequestSender, Stoppable {
+
+    private static final Logger LOG = LoggerFactory.getLogger(LwM2mRequestSenderImpl.class);
+    private final ExecutorService processingExecutor = Executors
+            .newCachedThreadPool(new NamedThreadFactory("leshan-response-processingExecutor-%d"));
+    private final ConcurrentLinkedQueue<ResponseListener> responseListeners = new ConcurrentLinkedQueue<ResponseListener>();
+
+    private final LwM2mRequestSender defaultRequestSender;
+    private final QueuedRequestSender queuedRequestSender;
+    private final ResponseListener queueModeResponseListener;
+    private final ResponseListener defaultResponseListener;
+    private final ClientRegistry clientRegistry;
+
+    /**
+     * default constructor.
+     * 
+     * @param defaultRequestSender used to send messages to a LWM2M Client using a non-queue mode.
+     * @param queuedRequestSender used to send messages to a LWM2M Client using a queue-mode.
+     */
+    public LwM2mRequestSenderImpl(final LwM2mRequestSender defaultRequestSender,
+            final QueuedRequestSender queuedRequestSender, final ClientRegistry clientRegistry) {
+        this.defaultRequestSender = defaultRequestSender;
+        this.queuedRequestSender = queuedRequestSender;
+        this.clientRegistry = clientRegistry;
+        this.queueModeResponseListener = queuedRequestSender.createResponseListener();
+        this.defaultResponseListener = createBindingAwareResponseListener();
+        this.defaultRequestSender.addResponseListener(defaultResponseListener);
+    }
+
+    @SuppressWarnings("deprecation")
+    @Override
+    public <T extends LwM2mResponse> T send(final Client registrationInfo, final DownlinkRequest<T> request,
+            final Long timeout) throws InterruptedException {
+        return defaultRequestSender.send(registrationInfo, request, timeout);
+    }
+
+    @SuppressWarnings("deprecation")
+    @Override
+    public <T extends LwM2mResponse> void send(final Client registrationInfo, final DownlinkRequest<T> request,
+            final ResponseCallback<T> responseCallback, final ErrorCallback errorCallback) {
+        if (registrationInfo.getBindingMode().equals(BindingMode.UQ)) {
```

Comment:
Introduce a `Client.usesQueueMode()` method or something like that instead of traversing the inner structure of Client...